### DISCUSSION TITLE
Parent spawning parallel state machines?

### DISCUSSION THREAD

**Discussion by victorias** [statelyai/xstate#2171] [Q&A]
Hi everyone! First of all, thank you so much @davidkpiano for this incredible library. And I also thank you (and this community) for being so kind and helpful when open source can be such a thankless task!

I'm trying to design a Form system using XState. Conceptually, I have several form Form Controls (ie, a Free Response Input, a Select field, etc), all of which has their own state machine managing their own internal states (such as active, pending, validating, etc). What I want to do is have a `FormStateMachine` that spawns each of the individual controls as child state machines (via `src`) and wait for each child to reach their final state. While designing it, I realized that maybe I don't fully understand how this is supposed to work. Here's some example code (and I'll elaborate later):

```ts
const createFormMachine = (childMachines: Array<StateMachine>) =>
    createMachine({
        id: "formMachine",
        initial: "active",
        context: {
            ...childMachines.reduce((prev, curr, idx) => {
                prev[`formcontrol-${idx}`] = false; // stores whether or not this child machine has finished
                return prev;
            }, {}),
        },
        states: {
            active: {
                invoke: childMachines.map((childMachine) => ({
                    src: childMachine,
                    onDone: () => null, // update context with child machine state, if all machines are done, transition
                })),
            },
        },
    });
```

This design stores state in context.. which definitely doesn't feel right! 😂  It seems like [parallel states](https://xstate.js.org/docs/guides/final.html#parallel-states) might be more appropriate (where each field is a parallel state?) but I really want to get a second opinion. Would I create parallel states for each Form Control, and thru each invoke its specific machine? Does spawning child state machines even make sense for forms, where "finality" is somewhat ambiguous?

Thanks!

**Comment by davidkpiano** [comment]
We have to improve guidance around forms, which can be tricky. In my opinion, each form field should be its own spawned actor. This makes things simpler since each field can have a "mind of its own" and communicate relevant changes to the parent form.

**Reply by ChrisShank** [reply]
If each form field is a spawned actor how would the parent machine keep track of which fields are valid so that the form can be submitted only if all the fields are valid?

**Reply by Andarist** [reply]
If you let the parent know about the change in a child - then sure, it could compute that information.

**Reply by ChrisShank** [reply]
Thanks! So I guess each child keeps track of whether it is valid and the parent has to also keep track of whether each child is valid?

**Reply by Andarist** [reply]
Yes, the important thing to reealize here is that actor's state is private so you can't just read it from the outside (in the pure actor model). This is important for backend frameworks because actors can live anywhere - in another thread or on another machine.

**Reply by ChrisShank** [reply]
Ya i’m still grappling with how much that goes against my intuition haha!

I was just reading this excerpt that sort of make it click for me

> The Erlang researchers initially believed that shared-memory was preventing fault-tolerance and they saw message-passing of immutable data between processes as the solution to avoiding shared-memory. There was a concern that passing around and copying data would be costly, but the Erlang developers saw fault-tolerance as a more important concern than performance.

http://dist-prog-book.com/chapter/3/message-passing.html

**Comment by Andarist** [comment]
I don't have strong opinions about forms with XState at the moment so I would like to refrain from giving specific guidance for the problem at hand.

But I can answer some question asked here:
> Does spawning child state machines even make sense for forms, where "finality" is somewhat ambiguous?

It makes sense. Child actors very often have "final" states and are only invoked/spawned between 2 time-points in the parent's lifetime but it also makes sense for a lot of situations to have them completely bound to the parent's lifetime (for them to be invoked/spawned right away and only being stopped when their parent stops). Think of it as organizational units - if you want to split out the children's logic out of a parent for better modularity etc then just go for it :) TLDR: children don't have to contain a final state
